test/dlc_tests/conftest.py

```diff
@@ -241,40 +241,61 @@ def ec2_public_ip(request):
     return request.param
 
 
-@pytest.fixture(scope="session")
-def region():
-    return os.getenv("AWS_REGION", DEFAULT_REGION)
+@pytest.fixture(scope="function")
+def region(request):
+    return request.param if hasattr(request, "param") else os.getenv("AWS_REGION", DEFAULT_REGION)
 
 
-@pytest.fixture(scope="session")
-def docker_client(region):
-    test_utils.run_subprocess_cmd(
-        f"$(aws ecr get-login --no-include-email --region {region})",
-        failure="Failed to log into ECR.",
-    )
-    return docker.from_env()
+@pytest.fixture(scope="function")
+def availability_zone_options(ec2_client, ec2_instance_type, region):
+    """
+    Parametrize with a reduced list of availability zones for particular instance types for which
+    capacity has been reserved in that AZ. For other instance types, parametrize with list of all
+    AZs in the region.
+    :param ec2_client: boto3 Client for EC2
+    :param ec2_instance_type: str instance type for which AZs must be determined
+    :param region: str region in which instance must be created
+    :return: list of str AZ names
+    """
+    allowed_availability_zones = None
+    if ec2_instance_type in ["p4de.24xlarge"]:
```

Comment:
isnt the region to instance type configuration already validated?

Reply:
We only want to use hardcoded availability-zone lists if the instance type being launched and the region both match pre-existing capacity reservations.

While the helper function to assign `instance-type + region` always associates only the region with capacity-reservations to the instance-type for now, we might have changes in the future where we may try to launch instances in other regions.

That's why we use the hardcoded AZ lists in specific scenarios, and get AZ IDs using the helper function otherwise.

Reply:
ack